### ISSUE TITLE
[Bug] Infinite loop in skmultilearn.cluster.random.RandomLabelSpaceClusterer.fit_predict(...)

### ISSUE THREAD

**Issue by andreplima** [scikit-multilearn-ng/scikit-multilearn-ng#24]
The fitting of a RakelO model to a multiclass dataset halts.

**To Reproduce**
﻿from sklearn.datasets import load_iris
from sklearn.preprocessing import LabelBinarizer
from skmultilearn.model_selection import iterative_train_test_split
from skmultilearn.ensemble import RakelO
from sklearn.ensemble import RandomForestClassifier

SEED=23
dataset = load_iris()
X = dataset.data
y = LabelBinarizer().fit(dataset.target).transform(dataset.target)
X_train, X_test, y_train, y_test = iterative_train_test_split(X, y, test_size = 0.2, random_state=SEED)
classifier = RakelO(
    base_classifier=RandomForestClassifier(n_estimators=500, random_state=SEED),
    base_classifier_require_dense=[True, True],
    labelset_size=3,
    model_count=2*y.shape[1],
)
classifier.fit(X_train, y_train)
predictions = classifier.predict(X_test)
print(predictions)

**Environment**
Python 3.9.19
scikit-multilearn-ng 0.0.6

**Possible fix**
Introduced counter variable to limit the number of iterations.

**Comment by northern-64bit** [comment]
Hello @andreplima !
Thanks for opening the issue.

After reviewing your example and the RakelO configuration, it appears that the significant delay in model fitting stems mainly from the high computational demands of the initially chosen parameters.

You use a large number of estimators (500) in the RandomForestClassifier. This, combined with a high model_count, resulted in a model that is computationally expensive and slow to train, especially given the relatively simple structure of the Iris dataset.

In order to fix this, I have tested a simple configuration that significantly reduces the training time (although the results are bad). Here’s the revised approach:

1. Decreasing the number of estimators in RandomForestClassifier to 2 reduces the complexity and computational load.
2. Lowering the model_count to 2 also helps in reducing the training time by decreasing the number of classifiers that need to be trained.
3. Setting labelset_size to 1 simplifies each classifier’s task, making the training process quicker.

Here’s the modified code that incorporates these optimizations:

```python
  from sklearn.datasets import load_iris
  from sklearn.preprocessing import LabelBinarizer
  from skmultilearn.model_selection import iterative_train_test_split
  from skmultilearn.ensemble import RakelO
  from sklearn.ensemble import RandomForestClassifier
  
  # Seed for reproducibility
  SEED = 23
  
  # Load the Iris dataset
  dataset = load_iris()
  X = dataset.data
  y = LabelBinarizer().fit_transform(dataset.target)
  
  # Split the data into training and testing sets
  X_train, X_test, y_train, y_test = iterative_train_test_split(X, y, test_size=0.2, random_state=SEED)
  
  # Initialize RakelO with a RandomForest classifier
  classifier = RakelO(
      base_classifier=RandomForestClassifier(n_estimators=2, random_state=SEED),
      base_classifier_require_dense=[True, True],
      labelset_size=1,
      model_count=2
  )
  
  # Fit the model on the training data
  classifier.fit(X_train, y_train)
  
  # Make predictions on the test data
  predictions = classifier.predict(X_test)
  
  # Print the predictions
  print(predictions)
```
This took way less time to run and finished quickly. If this doesn't work please reopen the issue.

Result:
```
  (0, 0)        1
  (1, 0)        1
  (2, 0)        1
  (3, 0)        1
  (4, 0)        1
  (5, 0)        1
  (6, 0)        1
  (7, 0)        1
  (8, 0)        1
  (9, 0)        1
  (10, 0)       1
  (11, 0)       1
  (12, 0)       1
  (13, 0)       1
  (14, 0)       1
  (15, 0)       1
  (16, 0)       1
  (17, 0)       1
  (18, 0)       1
  (19, 0)       1
  (20, 0)       1
  (21, 0)       1
  (22, 0)       1
  (23, 0)       1
  (24, 0)       1
  (25, 0)       1
  (26, 1)       1
  (27, 1)       1
  (28, 1)       1
  (29, 1)       1
```
(Btw. I wouldn't recommend the usage of RakelO for this dataset.)

Please try it out, and let me know if you encounter further issues or have any questions!
